chain/consensus/hierarchical/subnet/manager/crossmsg.go

```diff
@@ -0,0 +1,307 @@
+package subnetmgr
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/lotus/blockstore"
+	"github.com/filecoin-project/lotus/chain/actors"
+	"github.com/filecoin-project/lotus/chain/consensus/actors/registry"
+	"github.com/filecoin-project/lotus/chain/consensus/common"
+	"github.com/filecoin-project/lotus/chain/consensus/hierarchical"
+	"github.com/filecoin-project/lotus/chain/consensus/hierarchical/actors/sca"
+	"github.com/filecoin-project/lotus/chain/rand"
+	"github.com/filecoin-project/lotus/chain/stmgr"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/chain/vm"
+	"github.com/filecoin-project/lotus/node/modules/dtypes"
+	"github.com/filecoin-project/specs-actors/actors/util/adt"
+	"github.com/ipfs/go-cid"
+	cbor "github.com/ipfs/go-ipld-cbor"
+	"golang.org/x/xerrors"
+)
+
+// finalityWait is the number of epochs that we will wait
+// before being able to re-propose a cross-msg. This is used to
+// wait for all the state changes to be propagated.
+const finalityWait = 5
+
+func newCrossMsgPool() *crossMsgPool {
+	return &crossMsgPool{pool: make(map[hierarchical.SubnetID]*lastApplied)}
+}
+
+type crossMsgPool struct {
+	lk   sync.RWMutex
+	pool map[hierarchical.SubnetID]*lastApplied
+}
+
+type lastApplied struct {
+	lk      sync.RWMutex
+	topdown map[uint64]abi.ChainEpoch // nonce[epoch]
+	downtop map[uint64]abi.ChainEpoch // nonce[epoch]
+	height  abi.ChainEpoch
+}
+
+func (cm *crossMsgPool) getPool(id hierarchical.SubnetID, height abi.ChainEpoch) *lastApplied {
```

Comment:
I'd expect different pools for every relationship (one for parent, one each for children)

Reply:
I don't think this is needed. The transactions picked up in the pool have already been committed and checked by the SCA. The SCA aggregates and commits all transactions from childs and assigns a unique nonce to all of them. This pool picks up all of these transactions and apply them in order in the corresponding network consensus.

This is also related with you comment above. In any case, I definitely get the question because this PR is missing the code for the SCA. I really need to update the spec to have all the design in the same place. I'll ping you there when it is ready. Thanks!